core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -749,8 +749,10 @@ private[spark] class ExecutorAllocationManager(
           stageAttemptToNumRunningTask.getOrElse(stageAttempt, 0) + 1
         // If this is the last pending task, mark the scheduler queue as empty
         if (taskStart.taskInfo.speculative) {
-          stageAttemptToSpeculativeTaskIndices.getOrElseUpdate(stageAttempt,
-            new mutable.HashSet[Int]) += taskIndex
+          stageAttemptToSpeculativeTaskIndices
```

Comment:
Is there a concurrent access issue? `mutable.HashMap` is not thread safe

Reply:
`allocationManager.synchronized` will be called to protect this

Reply:
Got